pkg/api/pluginproxy/ds_auth_provider.go

```diff
@@ -24,7 +24,7 @@ type DSInfo struct {
 func ApplyRoute(ctx context.Context, req *http.Request, proxyPath string, route *plugins.Route,
 	ds DSInfo, cfg *setting.Cfg) {
 	proxyPath = strings.TrimPrefix(proxyPath, route.Path)
-
+	req.URL.Path = proxyPath
```

Comment:
This is being set in line 51, I infer that you intent to cover the case in which the route URL is not set. From the description in #80856 I believe that your problem is that you are using 

```
  "routes": [
    {
      "path": "xHeaderKey",
      "url": "{{ .JsonData.dataSourceUrl }}",
      "headers": [
        {
          "name": "X-ClickHouse-User",
          "content": "{{ .JsonData.xHeaderUser }}"
        },
        {
          "name": "X-ClickHouse-Key",
          "content": "{{ .SecureJsonData.xHeaderKey }}"
        }
      ]
    }
  ]
```

In your plugin.json, when what the `DataSourceHttpSettings` component sets is `url`. If you want to keep using that component, you need to drop the usage of `dataSourceUrl` and use `url` instead. Then define the routes like this:

```
  "routes": [
    {
      "path": "xHeaderKey",
      "url": "{{ .JsonData.url }}",
      "headers": [
        {
          "name": "X-ClickHouse-User",
          "content": "{{ .JsonData.xHeaderUser }}"
        },
        {
          "name": "X-ClickHouse-Key",
          "content": "{{ .SecureJsonData.xHeaderKey }}"
        }
      ]
    }
  ]
```

Reply:
```
"url": "{{ .JsonData.url }}",
```
it doesnt' work

because plugin don't have `JsonData.url`

`DataSourceHttpSettings` 
doesn't store data in JsonData, this component store data directly to Settings

I tried and got error
```
 logger=data-proxy-log userId=0 orgId=1 uname= path=/api/datasources/proxy/uid/PFA2E00E697AAEA0A/xHeaderKey/ remote_addr=172.26.0.1 referer=http://localhost:3000/connections/datasources/edit/PFA2E00E697AAEA0A t=2024-01-19T18:05:35.584044597Z level=error msg="Proxy request failed" err="unsupported protocol scheme \"\""
```

because `JsonData.url` is empty

Reply:
@andresmgot 
I need apply trick with
```
"url": "{{ .JsonData.dataSourceUrl }}",
```

because if i remove "url" from routes

then line 51 will never execute

and i always will get /xHeaderKey` in my `req.URL.path`
instead of `/`

Reply:
@andresmgot any news from your side?

Reply:
Sorry, indeed `"{{ .JsonData.url }}"` won't work since the URL is not stored in the JsonData. I just have a minor request. Can you move this line to line 52, in a `else` block, adding a comment why this needs to be set?. Thanks

Reply:
@andresmgot done https://github.com/grafana/grafana/pull/80858/commits/a95f05c51e37d0b0a4122bd7d2975afdec5695d9

Reply:
Skipping the URL would be the suggested solution since the Grafana datasource proxy automatically forwards to the configured URL. What you possibly want to change is the path, either leaving it empty would forward any request path to configured URL or define one route per each request path you want to allow.

Forward everything:
```json
  "routes": [
    {
      "path": "",
      "headers": [
        {
          "name": "X-ClickHouse-User",
          "content": "{{ .JsonData.xHeaderUser }}"
        },
        {
          "name": "X-ClickHouse-Key",
          "content": "{{ .SecureJsonData.xHeaderKey }}"
        }
      ]
    }
  ]
```

Forward only to abc and def:
```json
  "routes": [
    {
      "path": "abc",
      "headers": [
        {
          "name": "X-ClickHouse-User",
          "content": "{{ .JsonData.xHeaderUser }}"
        },
        {
          "name": "X-ClickHouse-Key",
          "content": "{{ .SecureJsonData.xHeaderKey }}"
        }
      ]
    },
    {
      "path": "def",
      "headers": [
        {
          "name": "X-ClickHouse-User",
          "content": "{{ .JsonData.xHeaderUser }}"
        },
        {
          "name": "X-ClickHouse-Key",
          "content": "{{ .SecureJsonData.xHeaderKey }}"
        }
      ]
    }
  ]
```

Reply:
@marefr nobody used empty `url` in `routes`

I check every line here https://github.com/search?q=path%3Aplugin.json%20routes&type=code, before submit my path

and current behavior is inconsistent
you rewrite path if `url` not empty
but pass path as is if `url` not present

And my header authorization configuration is optional, I can't always pass headers only if it defined by user, so I need to add `/xHeaderKey`  and need to remove it to properly pass 
because my url is `http://host:port/` not `http://host:port/xHeaderKey`

Reply:
Okay maybe I was wrong about the empty path. But you should use the star `*` it seems and then everything that you provide after `xHeaderKey/`/replaced with the star will be appended to the url, e.g.

```json
  "routes": [
    {
      "path": "xHeaderKey/*",
      "url": "{{ .JsonData.url }}",
      "headers": [
        {
          "name": "X-ClickHouse-User",
          "content": "{{ .JsonData.xHeaderUser }}"
        },
        {
          "name": "X-ClickHouse-Key",
          "content": "{{ .SecureJsonData.xHeaderKey }}"
        }
      ]
    }
  ]
```

Reply:
I assume that won't work either, the problem is that `DataSourceHttpSettings` won't store the URL in `.JsonData` so `route.URL` won't be populated and the path won't be replaced. Another option that would work I think is to [expand `templateData`](https://github.com/grafana/grafana/blob/main/pkg/api/pluginproxy/ds_auth_provider.go#L28) with the URL information from the datasource so the route could be defined like:

```
  "routes": [
    {
      "path": "xHeaderKey",
      "url": "{{ .URL }}",
      "headers": [
        {
          "name": "X-ClickHouse-User",
          "content": "{{ .JsonData.xHeaderUser }}"
        },
        {
          "name": "X-ClickHouse-Key",
          "content": "{{ .SecureJsonData.xHeaderKey }}"
        }
      ]
    }
  ]
```

That will replace `route.URL` with the right info and pass the check (`if len(route.URL) > 0`). WDYT?

Reply:
If it's only `"url": "{{ .URL }}",` you shouldn't use the URL property. Or am I missing something?

Reply:
> I think is to [expand `templateData`](https://github.com/grafana/grafana/blob/main/pkg/api/pluginproxy/ds_auth_provider.go#L28) with the URL information from the datasource so the route could be defined like:

Guess we can still allow this, but for this authors problem sounds to me it shouldn't be needed :shrug:

Reply:
@marefr I explained in details what is the problem. 
here 
https://github.com/grafana/grafana/pull/80858#discussion_r1459472608 
and here
https://github.com/grafana/grafana/pull/80858#discussion_r1459485847

@andresmgot thanks for understanding, yes you are right, we have two solutions here
1) expand allowed data in template
2) just make behavior consistent and not depend on is `url` present or not. 

I choose simplest solution 2.

Reply:
> If it's only "url": "{{ .URL }}", you shouldn't use the URL property. Or am I missing something?
> Guess we can still allow this, but for this authors problem sounds to me it shouldn't be needed 🤷

The problem with that, is that when reading `route.URL`, it will be empty so [this condition](https://github.com/grafana/grafana/pull/80858/files#diff-57633ecc518c65cc3c5da3a328e70ef85744f07820869362399307f4bf3b931dL35) will be false and the [path won't be replaced](https://github.com/grafana/grafana/pull/80858/files#diff-57633ecc518c65cc3c5da3a328e70ef85744f07820869362399307f4bf3b931dR50) (removed in this case), which is what the plugin needs.

as far as I understand 2. implies a breaking change so it's not an option. The only way I see this working and without a breaking change is 1.

> 1. expand allowed data in template
> 2. just make behavior consistent and not depend on is `url` present or not.
> 
> I choose simplest solution 2.

2. Implies a breaking change so the only way forward I see is 1. but @marefr would need to confirm.

Reply:
@marefr could you confirm about variant 1 solution?